pyop2/codegen/rep2loopy.py

```diff
@@ -0,0 +1,756 @@
+                                 predicates=predicates,
+                                 id=id,
+                                 depends_on=depends_on, depends_on_is_final=True)
+
+
+@singledispatch
+def expression(expr, parameters):
+    raise AssertionError("Unhandled expression type '%s'" % type(expr))
+
+
+@expression.register(Index)
+def expression_index(expr, parameters):
+    name = expr.name
+    if name not in parameters.domains:
+        vars = isl.make_zero_and_vars([name])
+        zero = vars[0]
+        domain = (vars[name].ge_set(zero) & vars[name].lt_set(zero + expr.extent))
+        parameters.domains[name] = domain
+    return pym.Variable(name)
+
+
+@expression.register(FixedIndex)
+def expression_fixedindex(expr, parameters):
+    return expr.value
+
+
+@expression.register(RuntimeIndex)
+def expression_runtimeindex(expr, parameters):
+    @singledispatch
+    def translate(expr, vars):
+        raise AssertionError("Unhandled type '%s' in domain translation" % type(expr))
+
+    @translate.register(Sum)
+    def translate_sum(expr, vars):
+        return operator.add(*(translate(c, vars) for c in expr.children))
+
+    @translate.register(Argument)
+    def translate_argument(expr, vars):
+        expr = expression(expr, parameters)
+        return vars[expr.name]
+
+    @translate.register(Variable)
+    def translate_variable(expr, vars):
+        return vars[expr.name]
+
+    @translate.register(Zero)
+    def translate_zero(expr, vars):
+        assert expr.shape == ()
+        return vars[0]
+
+    @translate.register(LogicalAnd)
+    def translate_logicaland(expr, vars):
+        a, b = (translate(c, vars) for c in expr.children)
+        return a & b
+
+    @translate.register(Comparison)
+    def translate_comparison(expr, vars):
+        a, b = (translate(c, vars) for c in expr.children)
+        fn = {">": "gt_set",
+              ">=": "ge_set",
+              "==": "eq_set",
+              "!=": "ne_set",
+              "<": "lt_set",
+              "<=": "le_set"}[expr.operator]
+        return getattr(a, fn)(b)
+
+    name = expr.name
+    if name not in parameters.domains:
+        lo, hi, constraint = expr.children
+        params = list(v.name for v in traversal([lo, hi]) if isinstance(v, (Argument, Variable)))
+        vars = isl.make_zero_and_vars([name], params)
+        domain = (vars[name].ge_set(translate(lo, vars))
+                  & vars[name].lt_set(translate(hi, vars)))
+        parameters.domains[name] = domain
+        if constraint is not None:
+            parameters.assumptions[name] = translate(constraint, vars)
+    return pym.Variable(name)
+
+
+@expression.register(MultiIndex)
+def expression_multiindex(expr, parameters):
+    return tuple(expression(c, parameters) for c in expr.children)
+
+
+@expression.register(Extent)
+def expression_extent(expr, parameters):
+    # FIXME: There will be a symbolic representation of this.
+    multiindex, = expr.children
```

Comment:
What is the best way to handle this node. I use it because the C petsc interface needs to know how many values I am passing in. Right now, it is the footprint of the indices we pass. But if we enabled more complicated transformations, this value would now be wrong.

Reply:
Yes, I was thinking if we keep it symbolically, we might be able to do apply more transformations for opaque function calls in loopy. Probably a separate PR.

Reply:
Thanks.